### PR TITLE
Optional and Accesses

### PR DESCRIPTION
**Pull request by gokulvanan**
Fresh pull request after merging optionals and Accesses changes.

### CODE REVIEW

**Comment by santanusinha** [comment]
Why didn't travis get triggered? Can you check once?

**Comment by gokulvanan** [comment]
Not sure, closing and will reopen this pull request to see if triggers travis build.

**Comment by gokulvanan** [comment]
reopening pull request

**Comment by santanusinha** [comment]
It's probably an issue with travis intg. Looking into it.

----
Upset the established order, and everything becomes chaos. I'm an agent of
chaos. Oh and you know the thing about chaos, it's fair.

**Comment by santanusinha** [comment]
Apparently they have removed adming acces to the repo for me. So I can't do
squat...

----
Upset the established order, and everything becomes chaos. I'm an agent of
chaos. Oh and you know the thing about chaos, it's fair.

**Comment by gokulvanan** [comment]
will ask gaurav to see.. I am wondering if the issue could be because this pull is from my fork master rather than branch as in previous cases..

**Comment by santanusinha** [comment]
No it's probably not. The way I see it, it's not integrated in travis (the checkbox is not toggled) probably.

**Comment by ToOnlyGaurav** [comment]
Updated Setting. Testing Travis. (While closing and reopening)

**Comment by gokulvanan** [comment]
Thanks Gaurav for fixing Travis.. This one is interesting.. The test case works in Java8 with which I was running.. but breaks when I switch to Java7.. Will debug this and update here..

**Comment by gokulvanan** [comment]
closed till I resolve this

**Comment by gokulvanan** [comment]
Issue was with ordering of builders being different in Java7 and Java8. 
This is cause with order of Keys in Map used in building List<List<DataBuilderMeta>> are different in both. But this is expected , the real issue is #7  which breaks in between as result this test case failed with reordered keys in java7. Currently commenting the Assertion and leaving the test to report performance numbers.

File: src/main/java/com/flipkart/databuilderframework/engine/DataBuilderMetadataManager.java [hunk 1]
```diff
@@ -22,15 +22,21 @@
     private Map<String, DataBuilderMeta> meta = Maps.newHashMap();
     private Map<String, List<DataBuilderMeta>> producedToProducerMap = Maps.newHashMap();
     private Map<String, TreeSet<DataBuilderMeta>> consumesMeta = Maps.newHashMap();
-
+    private Map<String, TreeSet<DataBuilderMeta>> optionalsMeta = Maps.newHashMap();
+	private Map<String, TreeSet<DataBuilderMeta>> accessesMeta = Maps.newHashMap();
```
Comment (santanusinha) [inline src/main/java/com/flipkart/databuilderframework/engine/DataBuilderMetadataManager.java]: indentation

File: src/main/java/com/flipkart/databuilderframework/engine/DataBuilderMetadataManager.java [hunk 2]
```diff
@@ -22,15 +22,21 @@
...
+	
     private DataBuilderMetadataManager(Map<String, Class<? extends DataBuilder>> dataBuilders,
                                        Map<String, DataBuilderMeta> meta,
                                        Map<String, List<DataBuilderMeta>> producedToProducerMap,
-                                       Map<String, TreeSet<DataBuilderMeta>> consumesMeta) {
+                                       Map<String, TreeSet<DataBuilderMeta>> consumesMeta,
+                                       Map<String, TreeSet<DataBuilderMeta>> optionalsMeta,
+                           				Map<String, TreeSet<DataBuilderMeta>> accessesMeta) {
```
Comment (santanusinha) [inline src/main/java/com/flipkart/databuilderframework/engine/DataBuilderMetadataManager.java]: indentation

File: src/main/java/com/flipkart/databuilderframework/engine/DataBuilderMetadataManager.java [hunk 3]
```diff
@@ -22,15 +22,21 @@
...
         this.dataBuilders = dataBuilders;
         this.meta = meta;
         this.producedToProducerMap = producedToProducerMap;
         this.consumesMeta = consumesMeta;
+		this.optionalsMeta = optionalsMeta;
```
Comment (santanusinha) [inline src/main/java/com/flipkart/databuilderframework/engine/DataBuilderMetadataManager.java]: dude...

File: src/main/java/com/flipkart/databuilderframework/engine/ExecutionGraphGenerator.java [hunk 4]
```diff
@@ -140,9 +140,9 @@ private DependencyNode generateDependencyTree(final String data, DataFlow dataFl
             if(null == info.getBuilder()) {
                 info.setBuilder(dataBuilderMeta.getName());
             }
-            for(String consumes : dataBuilderMeta.getConsumes()) {
+            for(String consumes : dataBuilderMeta.getEffectiveConsumes()) {
                 if(routeMeta.isAlreadyOnOutgoingPath(data, consumes)) {
-                    logger.info(String.format("Loop detected: Path for %s already contains %s", consumes, data));
+                    logger.debug(String.format("Loop detected: Path for %s already contains %s", consumes, data));
```
Comment (santanusinha) [inline src/main/java/com/flipkart/databuilderframework/engine/ExecutionGraphGenerator.java]: Should we let this be an info? Not a major thing anyways...

**Pull request merged**